nvflare/fuel/f3/drivers/net_utils.py

```diff
@@ -0,0 +1,282 @@
+
+log = logging.getLogger(__name__)
+
+LO_PORT = 1025
+HI_PORT = 65535
+MAX_ITER_SIZE = 10
+RANDOM_TRIES = 20
+SECURE_SCHEMES = {"https", "wss", "grpcs", "stcp", "satcp"}
+
+
+def bool_value(value) -> bool:
+    """Convert value into bool"""
+    if isinstance(value, str):
+        return value.lower() in {"true", "yes", "y", "t", "1"}
+
+    return value
+
+
+def ssl_required(params: dict) -> bool:
+    """Check if SSL is required"""
+    scheme = params.get(DriverParams.SCHEME.value, None)
+    return scheme in SECURE_SCHEMES or bool_value(params.get(DriverParams.SECURE.value, False))
+
+
+def get_ssl_context(params: dict, ssl_server: bool) -> Optional[SSLContext]:
+    if not ssl_required(params):
+        return None
+
+    ca_path = params.get(DriverParams.CA_CERT.value)
+    if ssl_server:
+        cert_path = params.get(DriverParams.SERVER_CERT.value)
+        key_path = params.get(DriverParams.SERVER_KEY.value)
+    else:
+        cert_path = params.get(DriverParams.CLIENT_CERT.value)
+        key_path = params.get(DriverParams.CLIENT_KEY.value)
+
+    if not all([ca_path, cert_path, key_path]):
+        scheme = params.get(DriverParams.SCHEME.value, "Unknown")
+        role = "Server" if ssl_server else "Client"
+        raise CommError(CommError.BAD_CONFIG, f"{role} certificate parameters are missing for scheme {scheme}")
+
+    if ssl_server:
+        ctx = ssl.create_default_context(ssl.Purpose.CLIENT_AUTH)
+    else:
+        ctx = ssl.create_default_context()
+
+    # This feature is only supported on 3.7+
+    ctx.minimum_version = ssl.TLSVersion.TLSv1_2
+    ctx.verify_mode = ssl.CERT_REQUIRED
+    ctx.check_hostname = False
+    ctx.load_verify_locations(ca_path)
+    ctx.load_cert_chain(certfile=cert_path, keyfile=key_path)
+
+    return ctx
+
+
+def get_address(params: dict) -> str:
+    host = params.get(DriverParams.HOST.value, "0.0.0.0")
+    port = params.get(DriverParams.PORT.value, 0)
+
+    return f"{host}:{port}"
+
+
+def parse_port_range(entry: Any):
+
+    if isinstance(entry, int):
+        return range(entry, entry + 1)
+
+    parts = entry.split("-")
+    if len(parts) == 1:
+        num = int(parts[0])
+        return range(num, num + 1)
+    lo = int(parts[0]) if parts[0] else LO_PORT
+    hi = int(parts[1]) if parts[1] else HI_PORT
+    return range(lo, hi + 1)
+
+
+def parse_port_list(ranges: Any) -> list:
+    all_ranges = []
+    if isinstance(ranges, list):
+        for r in ranges:
+            all_ranges.append(parse_port_range(r))
+    else:
+        all_ranges.append(parse_port_range(ranges))
+
+    return all_ranges
+
+
+def check_tcp_port(port) -> bool:
+    result = False
+    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+    try:
+        s.bind(("", port))
+        result = True
+    except Exception as e:
+        log.debug(f"Port {port} binding error: {e}")
+    finally:
+        s.close()
+
+    return result
+
+
+def get_open_tcp_port(resources: dict) -> Optional[int]:
```

Comment:
This is similar as the fuel.network_utils.get_open_ports(). Maybe we can combine these two later?

Reply:
That one doesn't support port ranges.

Reply:
Our plan is to drop the other one once the merge is done since no one else should be opening port other than FCI.